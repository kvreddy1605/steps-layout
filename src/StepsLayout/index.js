import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './index.scss';

class StepsLayout extends Component {
    constructor(props) {
        super(props);

        this.state = {
            currentComponent: this.props.startAtComponent,
            navState: this.getNavStates(this.props.startAtComponent, this.props.components.length),
        }

        this.jumpToComponent = this.jumpToComponent.bind(this);
        this.moveToNextComponent = this.moveToNextComponent.bind(this);
        this.moveToPreviousComponent = this.moveToPreviousComponent.bind(this);
    }
    
    jumpToComponent(index){
        const calNextComponentIndex = this.state.currentComponent + index;
        const lastComponentIndex = this.props.components.length - 1;

        if(calNextComponentIndex >= 0 && calNextComponentIndex <= lastComponentIndex  )
            this.setState(prevState => {   
                    return { currentComponent: calNextComponentIndex }
                },()=>this.setState(() => {
                    return {navState: this.getNavStates(this.state.currentComponent, this.props.components.length)}
                }
            )
        ); 
    }
    moveToNextComponent() {
        this.jumpToComponent(1);
    }
    moveToPreviousComponent() {
        this.jumpToComponent(-1);
    }
    
    renderProgressSteps() {
        const stepWidth = (100 / this.props.components.length ).toFixed(2) + "%";
        const style = { width : stepWidth }
        return this.props.components.map((s, i) => (
            <li className={this.getClassName("progress", i)} style={style} key={i} value={i}>
                {this.props.components[i].title}
            </li>
        ));
    }

    getNavStates(indx, length) {
        let styles = [];

        for (let i = 0; i < length; i++) {
            if (i < indx) {
                styles.push('done');
            }
            else if (i === indx) {
                styles.push('doing');
            }
            else {
                styles.push('todo');
            }
        }
        return styles 
    }

    getClassName(className, i) {
        let liClassName = className + "-" + this.state.navState[i];
        return liClassName;
    }

    render() {
        let compToRender;
        let cloneExtensions = {
            moveToNextComponent: () => {
                this.moveToNextComponent();
            },
            moveToPreviousComponent: () => {
                this.moveToPreviousComponent();
            },
            jumpToComponent: (incComponentIndex) => {
                this.jumpToComponent(incComponentIndex);
            },
        };
        const componentPointer = this.props.components[this.state.currentComponent].component;
        if (componentPointer instanceof Component || (componentPointer.type && componentPointer.type.prototype instanceof Component)) {
            cloneExtensions.ref = 'activeComponent';
        }
        compToRender = React.cloneElement(componentPointer, cloneExtensions);
        
        return (
            <div className="steps-layout">
                <ol className="steps-layout__progress">
                    {this.renderProgressSteps()}
                </ol> 
                <div className="steps-layout__component">         
                    {compToRender}
                </div>
            </div>
        );
    }
}


StepsLayout.propTypes = {
    components: PropTypes.array.isRequired,
    startAtComponent: PropTypes.number
}

StepsLayout.defaultProps = {
    startAtComponent: 0
}
export default StepsLayout;