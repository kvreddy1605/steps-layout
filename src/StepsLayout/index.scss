.steps-layout{
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  
  &__progress {
    width:100%;
    // max-width: 1000px;
    list-style-type: none;
    display: flex;

    li {
      position: relative;
      text-align: center; 

      &::after{
        content: '';
        position: absolute;
        width:100%;
        height: 22px;
        background: #ddd;
        top: 0px;
        z-index: -1;
      }
      
      @media (min-width: 768px) {
        &::after{
          left: -50%;
        }
        &::before {
          content: "1";
          width: 20px;
          height: 20px;
          border: 1px solid black;
          display: block;
          margin: 0 auto 10px auto;
          border-radius: 50%;
          background: #ddd;
          color: #ddd;
          text-align: center;
          font-weight: bold;
        }
        &:first-child::after{
          content: none;
        }
        &.progress-doing::after{
          background: #316BB6;
        }
        &.progress-done::after{
          background: #316BB6;
        }
        &.progress-doing::before{
          border-color: #316BB6;
          background: white;
          color: white;
        }
        &.progress-done::before{
          border-color: white;
          background: #316BB6;
          color: #316BB6;
        }
    
      }

      @media (max-width: 768px) { 

        &::after{
          height: 3px;
          top: 0px;
          left: 0px;
        }

        &::before {
          content: ">";
          position: absolute;
          left: 0%
        }
        &:first-child::before{
          content: none;
        }
        &.progress-doing::after{
          height: 10px;
          top: -4px;
          background: #666;
        }
      }
    }
  }

  @media (min-width: 768px) {
    .steps-layout__progress {
      padding:0px;
    }
  }
  @media (max-width: 768px) {
    .steps-layout__progress {
      font-size: 14px;
      padding: 10px 50px;

      li {
        padding: 10px;
      }
    }
  }

  &__component {
    width: 100%;
    text-align: center;
  }
}