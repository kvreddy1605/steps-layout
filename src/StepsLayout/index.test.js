import React from 'react';
import StepsLayout from './index';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

describe('StepsLayout Component Test', () => {

    it('SnapShot Test', () => {
        const wrapper = shallow(<StepsLayout components={[{title: 'hello', component: <p>hello</p>}]} />) 
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    it('On component initialization, it creates state currentComponent as expected when startAtComponent is not passed as props', () => {
        const wrapper = shallow(<StepsLayout components={[{title: 'hello', component: <p>hello</p>}]} />)
        expect(wrapper.state('currentComponent')).toBe(0);
    })
    it('On component initialization, it creates state navState as expected when startAtComponent is not passed as props', () => {
        const wrapper = shallow(<StepsLayout components={[{title: 'hello', component: <p>hello</p>}]} />)
        expect(wrapper.state('navState')).toEqual(['doing'])
    })
    it('On component initialization, it creates state currentComponent as expected when startAtComponent is passed as props', () => {
        const components = [
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>}
        ]
        const wrapper = shallow(<StepsLayout components={components} startAtComponent={1} />)
        expect(wrapper.state('currentComponent')).toBe(1)
    })
    it('On component initialization, it creates state navState as expected when startAtComponent is passed as props', () => {
        const components = [
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>}
        ]
        const wrapper = shallow(<StepsLayout components={components} startAtComponent={1} />)
        expect(wrapper.state('navState')).toEqual(['done', 'doing', 'todo'])
    })
    it('getNavStates method returns correctly', () => {
        const components = [
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>}
        ]
        const wrapper = shallow(<StepsLayout components={components} />)
        const actualValues = wrapper.instance().getNavStates(1, 3);
        const expectedValues = ['done', 'doing', 'todo'];
        expect(actualValues).toEqual(expectedValues);
    })
    it('classNames are assigned as expected', () => {
        const components = [
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>},
            {title: 'hello', component: <p>hello</p>}
        ]
        const wrapper = shallow(<StepsLayout components={components} />)
        wrapper.setState({ navState: ['done', 'doing', 'todo'] });
        expect(wrapper.find('.progress-doing').length).toBe(1);
        expect(wrapper.find('.progress-done').length).toBe(1);
        expect(wrapper.find('.progress-todo').length).toBe(1);
    })
})

