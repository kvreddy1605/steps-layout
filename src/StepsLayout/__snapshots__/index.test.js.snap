// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepsLayout Component Test SnapShot Test 1`] = `
<div
  className="steps-layout"
>
  <ol
    className="steps-layout__progress"
  >
    <li
      className="progress-doing"
      key="0"
      style={
        Object {
          "width": "100.00%",
        }
      }
      value={0}
    >
      hello
    </li>
  </ol>
  <div
    className="steps-layout__component"
  >
    <p
      jumpToComponent={[Function]}
      moveToNextComponent={[Function]}
      moveToPreviousComponent={[Function]}
    >
      hello
    </p>
  </div>
</div>
`;
